service/history/tasks/key.go

```diff
@@ -86,6 +87,26 @@ func (left Key) CompareTo(right Key) int {
 	return 0
 }
 
+func (k Key) Prev() Key {
+	if k.TaskID == 0 {
+		if k.FireTime.UnixNano() == 0 {
+			panic("Key encountered negative underflow")
+		}
+		return NewKey(k.FireTime.Add(-time.Nanosecond), math.MaxInt64)
+	}
+	return NewKey(k.FireTime, k.TaskID-1)
+}
+
+func (k Key) Next() Key {
+	if k.TaskID == math.MaxInt64 {
+		if k.FireTime.UnixNano() == math.MaxInt64 {
+			panic("Key encountered positive overflow")
+		}
+		return NewKey(k.FireTime.Add(time.Nanosecond), 0)
```

Comment:
do you plan to iterate by 1ns at a time for timer queue range?

Reply:
no, we will never reach here. If so, it means the taskID for this shard is already max int64 and the shard can't be used anymore. 

So yes we technically don't need this logic. But Wenquan suggests that this make more sense logically if you consider fire time as the high 64 bits of the task key. And when the lower 64bits (taskID) overflows, we add one to the fire time.